Python/compile.c

```diff
@@ -5696,15 +5696,11 @@ compiler_comprehension(struct compiler *c, expr_ty e, int type,
     }
     Py_CLEAR(entry);
 
-    if (is_async_generator && type != COMP_GENEXP &&
-        scope_type != COMPILER_SCOPE_ASYNC_FUNCTION &&
-        scope_type != COMPILER_SCOPE_COMPREHENSION &&
-        !is_top_level_await)
-    {
-        compiler_error(c, loc, "asynchronous comprehension outside of "
-                               "an asynchronous function");
-        goto error_in_scope;
-    }
+    assert (!is_async_comprehension ||
+            type == COMP_GENEXP ||
+            scope_type == COMPILER_SCOPE_ASYNC_FUNCTION ||
+            scope_type == COMPILER_SCOPE_COMPREHENSION ||
+            is_top_level_await);
```

Comment:
To avoid the compiler warning, maybe you can use:

```suggestion
            c->u->u_scope_type == COMPILER_SCOPE_ASYNC_FUNCTION ||
            c->u->u_scope_type == COMPILER_SCOPE_COMPREHENSION ||
            !IS_TOP_LEVEL_AWAIT(c));
```

and remove the temporary variables `scope_type` and `is_top_level_await` (now, those variables are only used in this assert according to GH). Note that the other assertion also uses `c->u->u_scope_type` and not the `scope_type` variable.

Reply:
I don't think your suggestion works because we need to read the variables before the `compiler_enter_scope()` call.

I might just remove the assert so we don't get a new compiler warning, let me think about it.

Reply:
Ah I didn't see the call to `compiler_enter_scope()`. Maybe instead of the assert, you could add a comment?

Reply:
(Just saw your `NDEBUG` guard)